File: src/test/java/cube8540/oauth/authentication/error/DefaultAuthenticationExceptionEntryPointTest.java
package cube8540.oauth.authentication.error;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.AuthenticationException;
import org.springframework.web.context.request.ServletWebRequest;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@DisplayName("OAuth2 보안 예외 처리 엔트리 클래스 테스트")
class DefaultAuthenticationExceptionEntryPointTest {

    private ExceptionTranslator<Object> translator;
    private ExceptionResponseRenderer<Object> renderer;

    private DefaultAuthenticationExceptionEntryPoint<Object> entryPoint;

    @BeforeEach
    @SuppressWarnings("unchecked")
    void setup() {
        this.translator = mock(ExceptionTranslator.class);
        this.renderer = mock(ExceptionResponseRenderer.class);

        this.entryPoint = new DefaultAuthenticationExceptionEntryPoint<>(translator, renderer);
    }

    @Test
    @DisplayName("응답 메시지 작성")
    @SuppressWarnings("unchecked")
    void renderingResponseMessage() throws Exception {
        ResponseEntity<Object> responseEntity = mock(ResponseEntity.class);
        AuthenticationException exception = mock(AuthenticationException.class);
        HttpServletRequest request = mock(HttpServletRequest.class);
        HttpServletResponse response = mock(HttpServletResponse.class);

        when(translator.translate(exception)).thenReturn(responseEntity);

        ArgumentCaptor<ServletWebRequest> requestCaptor = ArgumentCaptor.forClass(ServletWebRequest.class);
        entryPoint.commence(request, response, exception);
        verify(renderer, times(1)).rendering(eq(responseEntity), requestCaptor.capture());
        verify(response, times(1)).flushBuffer();
        assertEquals(request, requestCaptor.getValue().getRequest());
        assertEquals(response, requestCaptor.getValue().getResponse());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/error/DefaultAuthenticationExceptionEntryPointTest.kt
package cube8540.oauth.authentication.error

import io.mockk.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.http.ResponseEntity
import org.springframework.security.core.AuthenticationException
import org.springframework.web.context.request.ServletWebRequest
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class DefaultAuthenticationExceptionEntryPointTest {

    private val translator: ExceptionTranslator<Any> = mockk()
    private val renderer: ExceptionResponseRenderer<Any> = mockk()

    private val entryPoint = DefaultAuthenticationExceptionEntryPoint(translator, renderer)

    @Test
    fun `rendering response message`() {
        val requestCaptor = slot<ServletWebRequest>()

        val responseEntity: ResponseEntity<Any> = mockk(relaxed = true)
        val exception: AuthenticationException = mockk(relaxed = true)
        val request: HttpServletRequest = mockk(relaxed = true)
        val response: HttpServletResponse = mockk(relaxed = true)

        every { translator.translate(exception) } returns responseEntity
        every { renderer.rendering(eq(responseEntity), capture(requestCaptor)) } just Runs

        entryPoint.commence(request, response, exception)
        assertThat(requestCaptor.isCaptured).isTrue
        assertThat(requestCaptor.captured.request).isEqualTo(request)
        assertThat(requestCaptor.captured.response).isEqualTo(response)
        verifyOrder {
            renderer.rendering(responseEntity, requestCaptor.captured)
            response.flushBuffer()
        }
    }
}